keps/sig-api-machinery/3352-aggregated-discovery/README.md

```diff
@@ -0,0 +1,758 @@
+checking if there are objects with field X set) may be a last resort. Avoid
+logs or events for this purpose.
+-->
+
+###### How can someone using this feature know that it is working for their instance?
+
+<!--
+For instance, if this is a pod-related feature, it should be possible to determine if the feature is functioning properly
+for each individual pod.
+Pick one more of these and delete the rest.
+Please describe all items visible to end users below with sufficient detail so that they can verify correct enablement
+and operation of this feature.
+Recall that end users cannot usually observe component logs or access metrics.
+-->
+
+- [ ] Events
+  - Event Reason:
+- [ ] API .status
+  - Condition name:
+  - Other field:
+- [ ] Other (treat as last resort)
+  - Details:
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+<!--
+This is your opportunity to define what "normal" quality of service looks like
+for a feature.
+
+It's impossible to provide comprehensive guidance, but at the very
+high level (needs more precise definitions) those may be things like:
+  - per-day percentage of API calls finishing with 5XX errors <= 1%
+  - 99% percentile over day of absolute value from (job creation time minus expected
+    job creation time) for cron job <= 10%
+  - 99.9% of /health requests per day finish with 200 code
+
+These goals will help you determine what you need to measure (SLIs) in the next
+question.
+-->
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+<!--
+Pick one more of these and delete the rest.
+-->
+
+- [ ] Metrics
+  - Metric name:
+  - [Optional] Aggregation method:
+  - Components exposing the metric:
+- [ ] Other (treat as last resort)
+  - Details:
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+<!--
+Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
+implementation difficulties, etc.).
+-->
+
+### Dependencies
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### Does this feature depend on any specific services running in the cluster?
+
+<!--
+Think about both cluster-level services (e.g. metrics-server) as well
+as node-level agents (e.g. specific version of CRI). Focus on external or
+optional services that are needed. For example, if this feature depends on
+a cloud provider API, or upon an external software-defined storage or network
+control plane.
+
+For each of these, fill in the following—thinking about running existing user workloads
+and creating new ones, as well as about cluster-level services (e.g. DNS):
+  - [Dependency name]
+    - Usage description:
+      - Impact of its outage on the feature:
+      - Impact of its degraded performance or high-error rates on the feature:
+-->
+
+### Scalability
```

Comment:
Do you have an estimate as to what kind of performance increase we might see with this change? Is that something that you'd measure to determine how successful the new feature was (i.e. whether it was worth taking to beta etc)?

Per https://github.com/crossplane/crossplane/blob/60fc7d/design/one-pager-crd-scaling.md#client-side-throttling we currently see discovery taking ~20 seconds on EKS with ~360 fairly large API groups.

Reply:
(Turn off client side rate limiting and/or check your APF settings to solve in the mean time.) This should eliminate one source of that kind of problem, I can't promise you don't have others.

Reply:
@lavalamp I appreciate the input but it's not feasible for us as the maintainers of Crossplane to expect our community to go tune this. I believe even turning off client side rate limiting for discovery in kubectl (among other clients) currently requires forking - I don't think it's exposed as a flag.

I intuitively expect this KEP to result in much more performant discovery - what I'm curious about is whether you have any idea how much more performant it will be.

Reply:
@negz There are a variety of factors that affect the performance of discovery (latency between client-server, throughput, rate limiting, etc) so it's hard for us to give an exact answer. 

With this KEP, the total number of requests will be reduced from `kubectl api-versions | wc -l` to a single request. The total size of the content downloaded will be the exact same as the aggregate size of the current content downloaded via separate GVs (perhaps the aggregated document will be slightly smaller after introducing compression). Burst limits will no longer be a problem as only one request will be sent to fetch all the group-versions.